spring-cloud-data-core/src/main/java/org/springframework/cloud/data/core/StreamDefinition.java

```diff
@@ -16,11 +16,213 @@
+	 * This allows for retrieval of modules in the stream by index.
+	 * Modules are maintained in stream flow order (source is first, sink is last).
+	 *
+	 * @return list of module descriptors for this stream definition
+	 */
+	public List<ModuleDefinition> getModuleDefinitions() {
+		return Collections.unmodifiableList(modules);
+	}
+
+	/**
+	 * Return an iterator that indicates the order of module deployments for this
+	 * stream. The modules are returned in reverse order; i.e. the sink is returned
+	 * first followed by the processors in reverse order followed by the
+	 * source.
+	 *
+	 * @return iterator that iterates over the modules in deployment order
+	 */
+	public Iterator<ModuleDefinition> getDeploymentOrderIterator() {
+		return new ReadOnlyIterator<>(modules.descendingIterator());
+	}
+
+	@Override
+	public String toString() {
+		return new ToStringCreator(this)
+				.append("name", this.name)
+				.append(this.definition)
+				.toString();
+	}
+
+
+	/**
+	 * Builder object for {@link StreamDefinition} that supports
+	 * fluent style configuration.
+	 */
+	public static class Builder {
+
+		/**
+		 * @see StreamDefinition#name
+		 */
+		private String name;
+
+		/**
+		 * @see StreamDefinition#modules
+		 */
+		private List<ModuleDefinition> modules = new ArrayList<>();
+
+		/**
+		 * @see StreamDefinition#definition
+		 */
+		private String definition;
+
+		/**
+		 * Set the stream name
+		 *
+		 * @param name name of stream
+		 * @return this builder
+		 */
+		public Builder setName(String name) {
+			this.name = name;
+			return this;
+		}
+
+		/**
+		 * Set the DSL definition for the stream.
+		 *
+		 * @param definition DSL definition
+		 * @return this builder
+		 */
+		public Builder setDefinition(String definition) {
+			this.definition = definition;
+			return this;
+		}
+
+		/**
+		 * Add the provided {@link ModuleDefinition module definitions}
+		 * in the order provided by the list
+		 *
+		 * @param modules module definitions for stream
+		 * @return this builder
+		 */
+		public Builder addModuleDefinitions(List<ModuleDefinition> modules) {
+			this.modules.addAll(modules);
+			return this;
+		}
+
+		/**
+		 * Append a {@link ModuleDefinition module definition} to the
+		 * list of module definitions.
+		 *
+		 * @param module module definition to add to stream
+		 * @return this builder
+		 */
+		public Builder addModuleDefinition(ModuleDefinition module) {
+			this.modules.add(module);
+			return this;
+		}
+
+		/**
+		 * Create a new instance of {@link StreamDefinition}.
+		 *
+		 * @return new {@code StreamDefinition} instance
+		 */
+		public StreamDefinition build() {
+			return new StreamDefinition(this.name, this.definition, this.modules);
+		}
+	}
+
+
+	/**
+	 * Iterator that prevents mutation of its backing data structure.
+	 *
+	 * @param <T> the type of elements returned by this iterator
+	 */
+	private static class ReadOnlyIterator<T> implements Iterator<T> {
```

Comment:
Should we consider adding this to `CollectionUtils`? Guava also has this concept: 

http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/Iterators.html#unmodifiableIterator(java.util.Iterator)